### ISSUE TITLE
What causes multiple servers under 1 app pool?

### ISSUE THREAD

**Issue by ppumkin** [HangfireIO/Hangfire#457]
I noticed sometimes my production site will have multiple servers at various times. Mostly its one but sometimes two or three. I know its not really a problem but .. is it a misconfiguration on my side or could be a potential problem with something else.. or just normal, unavoidable behaviour as app pools restart?

![image](https://cloud.githubusercontent.com/assets/1249683/10245070/36a81664-68fc-11e5-8743-855402dc7394.png)

**Comment by mdymel** [comment]
I have the same issue. It keeps growing until we reset iis manually. It happens on different servers, one, two at a time: 
![image](https://cloud.githubusercontent.com/assets/418910/11266952/aa4551b4-8ea8-11e5-86f7-7f68b92f0c08.png)

**Comment by Sebazzz** [comment]
Same issue here, but a more extreme example. One instance is started 8 days ago, and one instance 2 days ago. Now we do happen to have a release around those days/time what caused the application pool to recycle. Would it be possible that some code of Hangfire is still lingering in memory and processing some jobs? We actually noticed because we store some information about the job in-memory (progress, mostly) and that progress information was not updated but via our logging we still found out that the job completed.

![meerdere-servers](https://cloud.githubusercontent.com/assets/1426097/15712764/d990b0da-2813-11e6-8ecc-90887dfb0c5f.png)

**Comment by odinserj** [comment]
Can you make a process dump and post here all the call stacks? You can do this via Task Manager → Details tab → Right-click on a process → Create dump file. This will generate a dump file that you can open with Visual Studio and see call stacks of all the threads. This gives almost immediate answer what is going on in your process.

**Comment by Sebazzz** [comment]
I still have the minidump, however, I cannot find the symbols for Hangfire so I won't be able to provide that.

**Comment by Sebazzz** [comment]
I have a minidump still, but I wasn't able to load all symbols. For what its worth, [the thread dump](http://pastebin.com/xtPMjs3z).

Also a screenshot I took earlier showing the two app domains:
![two-appdomains](https://cloud.githubusercontent.com/assets/1426097/16947553/0ee2d29c-4db0-11e6-9c21-f76478856745.png)

Powershell code:
```
cd IIS:\AppPools\nl_myapppool\WorkerProcesses
$c = Get-ChildItem
$a = $c.appDomains
$a.Collection 
```
This returns '1' showing that IIS has no knowledge of the second app domain.

**Comment by odinserj** [comment]
@Sebazzz, thanks a lot for the dump, it contains all the needed method names. Examining the contents...

**Comment by odinserj** [comment]
Whoa, 108 threads! @Sebazzz, would you like to send me the original `dmp` file via dropbox or something like that? You can send a secret link via email (available in my profile on GitHub).

**Comment by odinserj** [comment]
All the Hangfire threads are waiting for a shutdown cancellation token that is canceled during the application shutdown. So looks like something prevents your application from shutting down. **Hangfire itself can't do this**, because all of its threads [are background](https://msdn.microsoft.com/en-us/library/system.threading.thread.isbackground(v=vs.110).aspx).

Look at `Thread  89` on line 4573, there is a thread that processes an HTTP request to the Telerik's WebResource handler. It is very strange that there is a pending request to an appdomain that should not serve any request anymore. It should either be processed or timed out to allow appdomain to be unloaded.

There is no symbol information in call stack to analyze why the request is blocked, but we can at least investigate why it wasn't timed out, since pending request may prevent appdomain from being unloaded.

The request is stucked in the native [`MgdReadEntityBody`](http://referencesource.microsoft.com/#System.Web/Hosting/IISUnsafeMethods.cs,bcc8f06f73264af7) method call. If we go to the above method, [`ReadEntityCoreSync`](http://referencesource.microsoft.com/#System.Web/Hosting/IIS7WorkerRequest.cs,bea360c6267495f5) method, we'll see manipulations with the [`IsInReadEntitySync`](http://referencesource.microsoft.com/#System.Web/WorkerRequest.cs,c36adbc49ef81c4a,references) property with the following comment:

> // Mark a blocking call
>        // It allows RequestTimeoutManager to eventualy to close the connection and unblock the caller
>       // and handle request timeout properly **(if in cancelable state)**

So, the request should be in a *cancelable state* to allow RequestTimeoutManager to abort the corresponding thread. This may be the reason why it's still pending, so further investigation is required. This may be caused by ASP.NET itself, or Telerik's `WebResourceHandler`. To learn what's going on, we should start with the [`BeginCancelablePeriod`](http://referencesource.microsoft.com/#System.Web/HttpContext.cs,074771451fffbbbe) method call, since it affects the [`MustTimeout`](http://referencesource.microsoft.com/#System.Web/HttpContext.cs,acfadc2b5b8a1494,references) method call logic.

**Comment by ppumkin** [comment]
I think its more of an ASP.NET issue - I have been having these issues
since I reported it... they go away after a while so not really that
bothered by it. They get "cancelled" eventually - They just linger around
for a while. Everything still works as expected so that is the main
concern. This seems to be a very strange side effect.

**Comment by odinserj** [comment]
HTTP request is a subject to the `RequestTimeoutManager` logic only when both statements are true:

1. Execution timeout has a correct positive value (`<httpRuntime  executionTimeout="00:01:50"/>`)
2. Debugging mode is set to `false` (`<compilation debug="false">` tag).

It also looks like that long-running network requests may cause this, because *execution timeout* refers only to the amount of time the worker process spent processing the request; which is just a subset of time-taken (see [this SO answer](http://stackoverflow.com/a/4490897/1317575)).

**Comment by Sebazzz** [comment]
@odinserj I will send the minidump when I get to work tomorrow.

> All the Hangfire threads are waiting for a shutdown cancellation token that is canceled during the application shutdown. So looks like something prevents your application from shutting down. Hangfire itself can't do this, because all of its threads are background.

Yes, I thought so, but background threads don't matter for app domains do they? The app domain was recycled, but the application pool process doesn't necessarily need to be restarted for the new app domain to load.

**Comment by odinserj** [comment]
@Sebazzz, @ppumkin, @mdymel, how do you start the background processing, using the `UseHangfireServer`, `HangfireBootstrapper` (from making app always running) or by creating a custom `BackgroundJobServer` instance?

**Comment by odinserj** [comment]
@Sebazzz, in `Microsoft.Owin.Host.SystemWeb` source code I've found the following comment in the [`ShutdownDetector`](https://katanaproject.codeplex.com/SourceControl/latest#src/Microsoft.Owin.Host.SystemWeb/ShutdownDetector.cs) class:

> // Normally when the AppDomain shuts down IRegisteredObject.Stop gets called, except that
> // ASP.NET waits for requests to end before calling IRegisteredObject.Stop.

Unfortunately I can't prove this, because ASP.NET source code on referencesource is not enough, and I can't find when the [`PipelineRuntime.StopProcessing`](http://referencesource.microsoft.com/#System.Web/Hosting/IPipelineRuntime.cs,897508a6cd860128,references) method is called – it's backed by the `IPipelineRuntime` COM interface, and looks like it is called from IIS. So, let's trust ASP.NET guys, and keep in mind that `IRegisteredObject.Stop` method call may be postponed, or not called at all (at least there are some requests pending), like in your case.

Both `UseHangfireServer` and `HangfireBootstrapper` use `IRegisteredObject` interface to handle appdomain shutdown notifications from ASP.NET. Now we understand, that this is not enough to handle the edge cases. Despite of `UseHangfireServer` uses more aggressive checks provided by the `Microsoft.Owin.Host.SystemWeb`, it may not handle appdomain restarts caused by `bin` directory changes (like during the deployments).

So I'd like to add a new integration for ASP.NET applications that use all of these super-duper aggressive checks. Looks like it will use the reflection, but Katana developers already doing this in the `ShutdownDetector` class.

P.S. **Disable overlapped recycle**, works for *application POOL recycles*, but it doesn't work for *application DOMAIN restarts*. Configuration change and deployments do not cause *app pool recycle*, where a new worker is started. They only lead to appdomain restart. So this setting doesn't prevent your application from having two instances running.

P.P.S. Don't know, how exactly the `IIS:\AppPools\nl_myapppool\WorkerProcesses` resource works, but the [`RemoveThisAppDomainFromAppManagerTableOnce`](http://referencesource.microsoft.com/#System.Web/Hosting/HostingEnvironment.cs,009cca5ae4a6eb78,references) method is called before calling `IRegisteredObject.Stop`. If the resource is checking this appmanager tables, it may not count running yet app domains.

**Comment by Sebazzz** [comment]
@odinserj Using regular `UseHangfireServer` to setup Hangfire. Thumbs up for the investigation work!

**Comment by odinserj** [comment]
I've just released [Hangfire.AspNet 0.1.0](https://github.com/HangfireIO/Hangfire.AspNet) as a proof of concept (already available on NuGet). Together with `IRegisteredObject`, it uses timer-based checks for `HostingEnvironment.ShutdownReason` property. The value of this property is set before any calls to `IRegisteredObject.Stop` and any other blocking methods; and the value isn't reverted once it was set. So it is ideal candidate to check for.

Documentation is still pending, but you can use https://github.com/HangfireIO/Hangfire.Highlighter as an example.

**Comment by Serjster** [comment]
Following the Highlighter example to the T, deploying to my Azure App Service (S2, 2 Core 3.5 GB), keeps spawning more and servers, eventually bringing the App Service to its knees! Why does it keep spawning more servers?!? Why won't it stick to just 1 server! I have scoured the webs looking for a solution! This has been a very frustrating experience.

**Comment by samiraguiar** [comment]
I'm still having a similar problem, even after trying `Hangfire.AspNet`. Is there a way to overwrite the behavior of the `BackgroundProcessingServer` when [instantiating `BackgroundProcessContext`](https://github.com/HangfireIO/Hangfire/blob/6525aa314f65d1a029863928edf8c9cfc18ec30d/src/Hangfire.Core/Server/BackgroundProcessingServer.cs#L109) so that it doesn't use a Guid to identify the server? My idea is to have one server per application port.

**Comment by Zapnologica** [comment]
We are also experiencing issues with hangfire spawning many server instances, to the point when it brings our application pool to stop running.

Any idea how to fix this?  We been running fine for a good year but now recently it started doing this. only thing we have changes is adding queues to the hangfire setup.

**Comment by colotiline** [comment]
Same issue on .NET Core.

**Comment by Sebazzz** [comment]
@colotiline How are you hosting .NET Core? In-process, out-of-process IIS? Standalone? It seems very unlikely that this happens on .NET Core if the latter.

**Comment by colotiline** [comment]
@sebazzz 

Host:
Debug mode for a web application from VS code on a Windows desktop machine. 

Setup:
Standard setup from the documentation in the startup.cs. 

Database:
PostgreSQL 10. 

Issue:
After restarting the app I’ve got two servers of the hangfire.io.

**Comment by Sebazzz** [comment]
That is nowhere the same situation. This bug is about multiple instances after *proper* recycle of an application pool. This is problematic because the other instance is still active and also handles jobs (albeit with older versions of the assemblies if this happens after an redeploy). 

In your case the original instance is not active. It cannot be because the process that hosted it was terminated. 

Sidenote: for development purposes its better to include the HangFire.MemoryStorage package, especially in a multi-developer scenario when sharing the same database. If you would otherwise fire a job it might run at some other developer machine and you'Il have a hard time debugging.

**Comment by DaleCam** [comment]
If it helps anyone, we had this issue in Azure, and we figured out that there is a hangfire server per deployment slot, which makes logical sense. So stopping the deployment slots in the azure portal will stop the "extra" servers. I hope that helps!

**Comment by naza-b** [comment]
You are right @DaleCam but, what if I want all that slots and still don't want multiple hangfire server instances?

For that, you can add a different config file(or JSON config for .Net core) for each and every slot. after that you can validate if your environment is production or not and create a hangfire server instance like below.it will not create new instance until meet your condition and you can smoothly experience single instance of hangfire server.

![image](https://user-images.githubusercontent.com/70795139/109609744-c86baf80-7b51-11eb-83f7-cedda1c6e7c7.png)

also HangFire.MemoryStorage package suggestion from @Sebazzz  work for local debugging.
